MAINT: Enable E711 globally and autofix violations

saikiranp-24 opened the pull request:
MAINT: Enable E711 globally and autofix violations
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Enabled the flake8 E711 check (comparison to None) in `pyproject.toml` and fixed all existing violations globally across the codebase.

Changes:
- Removed `E711` from the ignore list in `pyproject.toml`.
- Ran `ruff check sympy --unsafe-fixes --fix` to replace all instances of `== None` and `!= None` with `is None` and `is not None`.

#### Other comments
**Context**
This global cleanup was suggested by @oscarbenjamin to permanently enforce singleton comparisons.

**Verification**
Verified locally by running the linter:
`ruff check sympy`
Result: All checks passed!

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
    * Enabled E711 check in `pyproject.toml` and fixed all violations (comparison to None should be `is None`).
<!-- END RELEASE NOTES -->

saikiranp-24 committed:
Fix: Replace '== None' with 'is None' in truss.py

saikiranp-24 commented:
Thanks for the suggestion @oscarbenjamin ! I agree, fixing it globally makes more sense. I will update pyproject.toml and apply the fixes now.

saikiranp-24 commented on sympy/codegen/ast.py:
```diff
@@ -202,7 +202,7 @@ def _get_constructor(cls, attr):
         """ Construct an attribute value from argument passed to ``__new__()``. """
         # arg may be ``NoneToken()``, so comparison is done using == instead of ``is`` operator
-        if arg == None:
+        if arg is None:
```
Thanks for pointing it out. I'm reverting this line to == None and adding a # noqa tag so the linter doesn't flag it again.

oscarbenjamin replied:
Read through the whole diff and review the changes carefully.

saikiranp-24 commented on sympy/printing/c.py:
```diff
@@ -506,7 +506,7 @@ def _print_Declaration(self, decl):
         else:
             raise NotImplementedError("Unknown type of var: %s" % type(var))
-        if val != None: # Must be "!= None", cannot be "is not None"
+        if val is not None: # Must be "!= None", cannot be "is not None"
```
I am currently reviewing all the changes made by executing ruff check --select E711 --fix . and reverting back to the original state wherever necessary. I will push an updated, safer version shortly. Thanks for your feedback and patience.

saikiranp-24 committed:
style: modernize None comparisons (E711)

Updated == None to 'is None' where appropriate. Preserved == None using # noqa in sensitive modules (printing, codegen, trigonometric) to maintain compatibility with custom sentinels and symbolic equality.

saikiranp-24 committed:
test: revert to symbolic equality in test_transforms.py

saikiranp-24 committed:
test: correct logic in test_issue_7181 (revert to != None)

oscarbenjamin commented on sympy/printing/c.py:
```diff
@@ -506,7 +506,7 @@ def _print_Declaration(self, decl):
         else:
             raise NotImplementedError("Unknown type of var: %s" % type(var))
-        if val != None: # Must be "!= None", cannot be "is not None"
+        if val != None:  # noqa: E711 # Must be "!= None", cannot be "is not None"
```
The comment here should be preserved as it explains why this is needed. Perhaps it should just be on the line above instead.

Same applies to all the other places.

saikiranp-24 committed:
style: use is None in codeprinter and move comments above noqa tags

saikiranp-24 commented:
Went through the error logs and noticed I left a trailing whitespace which led to as `AssertionError`. I'm currently away from my machine, I'll fix that shortly.

saikiranp-24 commented on sympy/printing/codeprinter.py:
Changed to `is None` here.
Preserved comments about sticking to `== None` or `!=None`  wherever they are present.
PEP8 E711 rule is now enabled globally, open to change something if required.
